### ISSUE TITLE
websocket server sends duplicate message headers in protos response

### ISSUE THREAD

**Issue by mjay-cr** [gazebosim/gz-launch#246]
## Environment
* OS Version: Ubuntu 20.04
Binary gz-harmonic package from ubuntu-stable
 

## Description
* Expected behavior:  Running shapes.sdf simulation, then websocket server and connecting with [web app](https://app.gazebosim.org/visualization) shows visualisation of simulation in browser.
* Actual behavior: Nothing happens when you click connect, looking at console log gives error.

## Steps to reproduce

1. Install gz-harmonic
2. On a terminal run  `gz sim -v 4 -s -r shapes.sdf`
3. In another terminal run `gz launch -v 4 /usr/share/gz/gz-launch7/configs/websocket.gzlaunch`
4. Go to [web app](https://app.gazebosim.org/visualization) and click connect

## Output
If you open up the dev-tools on the browswer and then look at console, you will see the error:
```
ERROR Error: duplicate name 'Any' in Namespace .gz.msgs
    add namespace.js:240
    parseType parse.js:354
    parseCommon parse.js:261
    parse parse.js:790
    onloadend gzweb.module.js:11673
```
Switching to network and filtering by WebSocket requests, you will see that gzweb sends protos request. (Looking at the gz_launch console log will also show that it has been received). Then the response, when viewed as UTF-8 will see the proto headers sent below. After Wrench message it seems to repeat? Response from gz-garden and gz-harmonic are attached. Note: that the web visualisation works with garden and fortress (as per the steps above)

[gz-garden-response.txt](https://github.com/gazebosim/gz-launch/files/14204855/gz-garden-response.txt)
[gz-harmonic-response.txt](https://github.com/gazebosim/gz-launch/files/14204857/gz-harmonic-response.txt)

Not sure if these duplicate headers are the problem but seems like a bug

**Comment by hughlv** [comment]
Before this issue gets solved, what is the best workaround? I need to run gazebo as a websocket server so I can use gzweb2 to render on webpage.

**Comment by VincidaB** [comment]
After some testing, I was able to see that `WebsocketServer.cc` seem to receive the types in double on this line :
```c++
 gz::msgs::Factory::Types(types);
 ```
This would make it an issue with `gz-msgs` probably.

I was able to come-up with a dirty fix on the `gz-launch` side : 
```c++
std::set<std::string> processedDescriptors;

// Get all the messages, and build a single proto to send to the client.
for (auto const &type : types)
{
  auto msg = gz::msgs::Factory::New(type);
  if (msg)
  {
    auto descriptor = msg->GetDescriptor();
    if (descriptor){
      std::string descriptorStr = descriptor->DebugString();
      std::string firstLine = descriptorStr.substr(0, descriptorStr.find('\n'));
      if (processedDescriptors.find(firstLine) != processedDescriptors.end())
      {
        gzdbg << "Skipping duplicate descriptor[" << firstLine << "]\n";
        continue; // Skip if the first line of the descriptor has already been processed
      }
      allProtos += descriptorStr;
      gzdbg << descriptorStr;
      processedDescriptors.insert(firstLine); // Mark this descriptor as processed
    }
    else
    {
      gzerr << "Failed to get the descriptor for message["
        << type << "]\n";
    }
  }
  else
  {
    gzerr << "Failed to build message[" << type << "].\n";
  }
}
```

**Comment by iche033** [comment]
I created a couple of PRs: 

* https://github.com/gazebosim/gz-launch/pull/281
* https://github.com/gazebosim/gz-msgs/issues/470

I think that may fix the issue. I was able to get shapes.sdf working after these changes.

**Comment by iche033** [comment]
Maybe are there are still system installed gz-msgs10 or gz-launch7 packages in the system? That error msg should be fixed in 
`gz-launch 7.1.1`. Maybe you still  `libgz-launch7-dev` 7.1.0 or `ros-jazzy-gz-launch-vendor` installed?

**Comment by annamannucci** [comment]
> Maybe are there are still system installed gz-msgs10 or gz-launch7 packages in the system? That error msg should be fixed in `gz-launch 7.1.1`, and building from source with latest `gz-launch7` branch should be good. Maybe you still `libgz-launch7-dev` 7.1.0 or `ros-jazzy-gz-launch-vendor` installed?

Yes. I'm using gz-harmonic in combination with ros-jazzy. To expose the launch command, I have to `sudo apt-get install ros-jazzy-gz-launch-vendor`. This seems to come with gz-launch 7.1.0 (see https://github.com/gazebo-release/gz_launch_vendor/blob/jazzy/CMakeLists.txt#L8) and thus leads to the error described above. Anyway, I see that there is an open PR to upgrade the library to 7.1.1: https://github.com/gazebo-release/gz_launch_vendor/pull/4. I'll then wait for this PR to be merged (hopefully in a few days) :)

Best,
Anna
